content/article/why-i-hate-json-and-you-can-too-.md

```diff
@@ -0,0 +1,177 @@
+
+  {
+    "title"       : "Why I Hate JSON (and you can, too)",
+    "authors"     : ["felipe-gasper"],
+    "date"        : "2019-11-10T21:52:51",
+    "tags"        : [],
+    "draft"       : true,
+    "image"       : "",
+    "thumbnail"   : "",
+    "description" : "Another look at this ubiquitous serialization …",
+    "categories"  : "data"
+  }
+
+[JSON](https://www.json.org/) needs no introduction among most programmers.
+Since its “discovery” by Douglas Crockford in the early 2000s it has become
+one of the most popular data serialization formats yet created, having
+effectively supplanted XML as the web’s “go-to” format for representing
+structured data in APIs.
+
+For all its popularity, though, JSON carries a number of liabilities that
+make it less than ideal in some contexts where it nonetheless enjoys
+widespread use. This article will discuss some of these and propose
+alternative formats that I think deserve your attention.
+
+
+# 1. UTF-8
+
+JSON’s chief liability is its UTF-8 restriction: all JSON documents must
+be UTF-8, from start to finish. This means any application that wants to
+exchange binary data either cannot use JSON, or—perhaps more commonly—uses
+a secondary encoding like Base64 to store binary strings as text.
+Such techniques carry performance penalties and eventually create consistency
+problems, such as in the
+[WAMP](https://wamp-proto.org/_static/gen/wamp_latest_ietf.html) protocol,
+where it’s impossible to distinguish an “encoded” binary
+string from a text string that happens to fit the pattern that the protocol
+recommends for encoding binary strings to JSON.
+
+UTF-8 is a fine way of encoding text, but applications that use JSON bind
+themselves to a restriction that may prove to have been imprudent—sometimes
```

Comment:
There's nothing the prevents an application from using other things in relationship to JSON. I think you're torturing the point a bit here.

Reply:
What I mean is that an API that embraces JSON as its serialization format is incapable of (natively) transmitting anything that isn’t UTF-8. So if your input is, e.g., a Windows-1252 email body, you’re screwed.

Reply:
But your point is not a problem with JSON. Again, UTF-8 is an encoding, not the data.

There are plenty of APIs that accept different encodings for input.

Again, I'm not trying to argue a point here, but these are the reactions you'll get from readers so it's your job to deal with those :)

Reply:
I hear ya Felipe: to use JSON means to require UTF-8 encoding. I wonder what the practical implications are? For instance if the Content-Type header for an HTTP request is:

    application/json; charset=UTF-16

Does that imply that something (webserver / middleware / web app) must transform the request body to UTF-8 before the JSON parser can evaluate it? That's interesting as if the application interpreter was UT8-16 native, it implies it must undergo additional encoding/decoding every time it encodes or decodes JSON. OTOH, I understand that this means the request body is not officially "JSON" but it does seem like HTTP is capable of transmitting JSON data in encodings other than UTF-8 too.

Regardless, this paragraph seems OK to me, if you're happy with it, you might get some nitpickers is all.

Reply:
JSON doesn't really require UTF-8. It's just that UTF-8 is the only UTF that can spot encoding errors such as a missing byte.

Reply:
RFC 8259 says:
```
8.1.  Character Encoding

   JSON text exchanged between systems that are not part of a closed
   ecosystem MUST be encoded using UTF-8 [RFC3629].

   Previous specifications of JSON have not required the use of UTF-8
   when transmitting JSON text.  However, the vast majority of JSON-
   based software implementations have chosen to use the UTF-8 encoding,
   to the extent that it is the only encoding that achieves
   interoperability.

   Implementations MUST NOT add a byte order mark (U+FEFF) to the
   beginning of a networked-transmitted JSON text.  In the interests of
   interoperability, implementations that parse JSON texts MAY ignore
   the presence of a byte order mark rather than treating it as an
   error.
```
^^ How is that not requiring UTF-8?

Reply:
Felipe, I'm only working as an editor here. If you say that JSON requires UTF-8, you have to explain where that's applicable. Your job as the author is to fill in all these things for the reader, so you should say in the text "unless you are working purely internally, RFC 8259 says you must use UTF-8".

Remember, most of the people (including me) have not thought through the problem as much as you have. As such, you need to fill in these details for the readers and ensure that what you say is true. JSON does not require UTF-8 (say you say), but interchange between different systems does. So, that's the claim you should make and emphasize.

If you have the source for something you want to claim, include that source in the article so that the reader knows more. And, you get more credibility this way. There are plenty of "I hate X" articles out there, but you can distinguish your by showing off your command of the material and standards.

For example, in [If you’re like me and HATE JSON…](https://medium.com/@cuttenweiler/if-youre-like-me-and-hate-json-20aadfb18254), he doesn't really have a gripe with JSON. It's the poor quality of the AWS docs that drives him mad.

Here's a better balanced article: [Why Do People Dislike JSON?](https://apisyouwonthate.com/blog/why-do-people-dislike-json). Same sort of clickbait title, but there's a quite a bit of historical perspective and a deeper dive into the revealed instead of stated problem. It acknowledges that there is human choice and decision involved with much of that pain that you are describing. 

[Why I hate “JSON API 1.0” – a Mobile Dev’s perspective](https://elliotschrock.com/2018/11/12/why-i-hate-json-api-1-0-a-mobile-devs-perspective/) starts by setting the stage and limiting its perspective, then shows a particular and well-articulated pain point. But, it doesn't go beyond that concrete pain: "That’s why I hate JSON API 1.0. It was optimized for problems I don’t have, and creates a ton more work (and cost!) by invalidating swathes of solutions for well known problems that I, as a mobile dev, do have."